common/http_proxy_test.cpp

```diff
@@ -0,0 +1,913 @@
+		req,
+		[&client_hit_header, &received](http::ExpectedIncomingResponsePtr exp_resp) {
+			ASSERT_TRUE(exp_resp) << exp_resp.error().String();
+			auto resp = exp_resp.value();
+			EXPECT_EQ(resp->GetStatusCode(), 200);
+			client_hit_header = true;
+
+			auto body_writer = make_shared<io::ByteWriter>(received);
+			body_writer->SetUnlimited(true);
+			resp->SetBodyWriter(body_writer);
+		},
+		[&client_hit_body, this](http::ExpectedIncomingResponsePtr exp_resp) {
+			ASSERT_TRUE(exp_resp) << exp_resp.error().String();
+			client_hit_body = true;
+			loop.Stop();
+		});
+	ASSERT_EQ(error::NoError, err);
+
+	loop.Run();
+
+	EXPECT_TRUE(plain_server_hit_header);
+	EXPECT_TRUE(plain_server_hit_body);
+	EXPECT_TRUE(client_hit_header);
+	EXPECT_TRUE(client_hit_body);
+	EXPECT_EQ(common::StringFromByteVector(received), "Test\r\n");
+}
+
+TEST_F(HttpProxyHttpTest, WrongProxySet) {
+	bool client_hit_header = false;
+
+	http::ClientConfig client_config {
+		.http_proxy = "http://127.0.0.1:1",
+	};
+	http::Client client(client_config, loop);
+	auto req = make_shared<http::OutgoingRequest>();
+	req->SetMethod(http::Method::GET);
+	req->SetAddress("http://127.0.0.1:" TEST_PORT "/index.html");
+	auto err = client.AsyncCall(
+		req,
+		[&client_hit_header, this](http::ExpectedIncomingResponsePtr exp_resp) {
+			EXPECT_FALSE(exp_resp);
+			client_hit_header = true;
+			loop.Stop();
+		},
+		[](http::ExpectedIncomingResponsePtr exp_resp) {
+			EXPECT_TRUE(false) << "Should not get here";
+		});
+	ASSERT_EQ(error::NoError, err);
+
+	loop.Run();
+
+	EXPECT_FALSE(plain_server_hit_header);
+	EXPECT_FALSE(plain_server_hit_body);
+	EXPECT_TRUE(client_hit_header);
+}
+
+TEST_F(HttpProxyHttpTest, BogusProxySet) {
+	auto test = [this](const string &protocol, const http::ClientConfig &client_config) {
+		http::Client client(client_config, loop);
+		auto req = make_shared<http::OutgoingRequest>();
+		req->SetMethod(http::Method::GET);
+		req->SetAddress(protocol + "://localhost/index.html");
+
+		auto handler = [](http::ExpectedIncomingResponsePtr exp_resp) {
+			EXPECT_TRUE(false) << "Should not get here";
+		};
+		return client.AsyncCall(req, handler, handler);
+	};
+
+	auto err = test("http", http::ClientConfig {.http_proxy = "bogus"});
+	EXPECT_NE(error::NoError, err);
+	EXPECT_EQ(err.code, http::MakeError(http::InvalidUrlError, "").code) << err.String();
+
+	err = test("http", http::ClientConfig {.http_proxy = "http://localhost/a-path"});
+	EXPECT_NE(error::NoError, err);
+	EXPECT_EQ(err.code, http::MakeError(http::InvalidUrlError, "").code) << err.String();
+
+	err = test("https", http::ClientConfig {.https_proxy = "bogus"});
+	EXPECT_NE(error::NoError, err);
+	EXPECT_EQ(err.code, http::MakeError(http::InvalidUrlError, "").code) << err.String();
+
+	err = test("https", http::ClientConfig {.https_proxy = "http://localhost/a-path"});
+	EXPECT_NE(error::NoError, err);
+	EXPECT_EQ(err.code, http::MakeError(http::InvalidUrlError, "").code) << err.String();
```

Comment:
We don't have any test setting `.no_proxy`, can we add one?

Reply:
No, because any value is valid there, at least the way we have implemented it. I guess you could say that `..myhost` is invalid, but without a proper standard it's hard to decide what to do with it other than treat it as a wildcard suffix with value `.myhost`.